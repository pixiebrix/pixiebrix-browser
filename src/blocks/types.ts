/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import {
  Expression,
  OutputKey,
  RegistryId,
  TemplateEngine,
  UUID,
} from "@/core";
import { UnknownObject } from "@/types";

/**
 * @see https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API
 * @since 1.4.10
 */
export type URLPattern = string | URLPatternInit;

/**
 * An availability rule. For a rule to match, there must be match from each of the provided entries.
 *
 * An empty value (null, empty array, etc.) matches any site.
 *
 * @see checkAvailable
 */
export type Availability = {
  /**
   * Used to request permissions from the browser
   */
  matchPatterns?: string | string[];
  /**
   * NOTE: the urlPatterns must be a subset of matchPatterns (i.e., more restrictive). If not, PixieBrix may not have
   * access to the page
   *
   * @see https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API
   * @since 1.4.10
   */
  urlPatterns?: URLPattern | URLPattern[];
  /**
   * A selector that must be available on the page in order for the extension to be run.
   *
   * NOTE: the selector must be available at the time the contentScript is installed. While the contentScript is loaded
   * on document_idle, for SPAs this may lead to races between the selector check and rendering of the front-end.
   */
  selectors?: string | string[];
};

/**
 * Availability with consistent shape (i.e., all fields are arrays if provided)
 * @see Availability
 */
export type NormalizedAvailability = {
  matchPatterns?: string[];
  urlPatterns?: URLPattern[];
  selectors?: string[];
};

export type ReaderConfig =
  | RegistryId
  | { [key: string]: ReaderConfig }
  | ReaderConfig[];

/**
 * Where to execute the brick
 * - self: the current tab
 * - opener: the tab that opened the current tab
 * - target: the last tab that the current tab opened
 * - top: the top-most frame in the window
 * - broadcast: all tabs that PixieBrix has access to (the result is returned as an array)
 * - remote: the server (currently only support identity, get, and http bricks)
 * @see {@link BlockConfig.window}
 */
export type BlockWindow =
  | "self"
  | "opener"
  | "target"
  | "top"
  | "broadcast"
  | "remote";

/**
 * Condition expression written in templateEngine for deciding if the step should be run.
 * @see {@link BlockConfig.if}
 */
export type BlockIf = string | boolean | number | Expression;

/**
 * A block configuration to be executed by the PixieBrix runtime.
 * @see runStage
 * @see reducePipeline
 */
export type BlockConfig = {
  /**
   * The registry id of the configured block
   */
  id: RegistryId;

  /**
   * The configuration of the block
   */
  config: UnknownObject;

  /**
   * (Optional) human-readable label for the step.
   *
   * Shown in the page editor, logs, and the progress indicator
   * @see notifyProgress
   */
  label?: string;

  /**
   * (Optional) `true` to indicate on the host page that the step is being run
   */
  notifyProgress?: boolean;

  /**
   * (Optional) If the brick is being run in the context of a deployment, if an error occurs in this step an
   * alert email is send to the admins for the deployment with the brick input.
   *
   * Used as a stopgap/recovery mechanism operations that aren't safe to retry. (For example, sending data to a
   * UiPath queue for execution)
   */
  onError?: {
    alert?: boolean;
  };

  /**
   * Where to execute the brick (default=`self`)
   * @see BlockWindow
   */
  window?: BlockWindow;

  /**
   * The output key (without the preceding "@") to assign the brick output to
   *
   * As of brick API `v2`, the outputKey is required except:
   * - Effect bricks (because they don't produce an output)
   * - The last brick in pipeline
   *
   * @see ApiVersionOptions.explicitDataFlow
   */
  outputKey?: OutputKey;

  /**
   * (Optional) condition expression written in templateEngine for deciding if the step should be run. If not
   * provided, the step is run unconditionally.
   * @see BlockIf
   */
  if?: BlockIf;

  /**
   * (Optional) whether the block should inherit the current root element, or if it should use the document
   * (default=`inherit`)
   *
   * @see root
   * @since 1.4.0
   */
  rootMode?: "inherit" | "document";

  /**
   * (Optional) root jQuery/CSS selector. The selector is relative to the `root` that is passed to the pipeline/stage.
   *
   * An error is thrown at runtime if the selector doesn't match exactly one element
   * @see rootMode
   */
  root?: string;

  /**
   * (Optional) template language to use for rendering the `if` and `config` properties (default=`mustache`)
   * @see config
   * @see if
   */
  templateEngine?: TemplateEngine;

  /**
   * A unique id for the configured block, used to correlate traces across runs when using the Page Editor.
   *
   * DO NOT SET: generated automatically by the Page Editor when configuring a dynamic element.
   */
  instanceId?: UUID;
};

/**
 * A pipeline of blocks to execute sequentially
 */
export type BlockPipeline = BlockConfig[];

/**
 * A control flow branch, for tracing. The array of branches is used to correlate runs of the same block in a pipeline
 * @see TraceMetadata.blockInstanceId
 * @see TraceMetadata.branches
 */
export type Branch = {
  key: string;
  counter: number;
};
