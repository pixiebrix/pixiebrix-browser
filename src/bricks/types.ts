/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { type RegistryId } from "@/types/registryTypes";
import {
  type ElementReference,
  type Expression,
  type OutputKey,
  type TemplateEngine,
} from "@/types/runtimeTypes";
import { type UUID } from "@/types/stringTypes";
import { type UnknownObject } from "@/types/objectTypes";

/**
 * @see https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API
 * @since 1.4.10
 */
type URLPattern = string | URLPatternInit;

/**
 * An availability rule. For a rule to match, there must be match from each of the provided entries.
 *
 * An empty value (null, empty array, etc.) matches any site.
 *
 * @see checkAvailable
 */
export type Availability = {
  /**
   * Used to request permissions from the browser
   */
  matchPatterns?: string | string[];
  /**
   * NOTE: the urlPatterns must be a subset of matchPatterns (i.e., more restrictive). If not, PixieBrix may not have
   * access to the page
   *
   * @see https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API
   * @since 1.4.10
   */
  urlPatterns?: URLPattern | URLPattern[];
  /**
   * A selector that must be available on the page in order for the extension to be run.
   *
   * NOTE: the selector must be available at the time the contentScript is installed. While the contentScript is loaded
   * on document_idle, for SPAs this may lead to races between the selector check and rendering of the front-end.
   */
  selectors?: string | string[];
};

/**
 * Availability with consistent shape (i.e., all fields are arrays if provided)
 * @see Availability
 */
export type NormalizedAvailability = {
  matchPatterns?: string[];
  urlPatterns?: URLPattern[];
  selectors?: string[];
};

export type ReaderConfig =
  | RegistryId
  | { [key: string]: ReaderConfig }
  | ReaderConfig[];

/**
 * Where to execute the brick
 * - self: the current tab/frame
 * - opener: the tab that opened the current tab
 * - target: the last tab that the current tab opened
 * - top: the top-most frame in the window
 * - all_frames: all frames in the window
 * - broadcast: all tabs that PixieBrix has access to (the result is returned as an array)
 * @see {@link BrickConfig.window}
 * @since 1.7.36 - added all_frames
 */
export type BrickWindow =
  | "self"
  | "opener"
  | "target"
  | "top"
  | "broadcast"
  | "all_frames";

/**
 * Returns true if the BrickWindow targets multiple tabs and/or frames.
 *
 * The runtime will return an array of results (one for each successful target)
 */
export function hasMultipleTargets(
  window: BrickWindow
): window is "broadcast" | "all_frames" {
  return window === "broadcast" || window === "all_frames";
}

/**
 * Condition expression written in templateEngine for deciding if the step should be run.
 * @see {@link BrickConfig.if}
 */
export type BlockIf = string | boolean | number | Expression;

/**
 * A brick configuration to be executed by the PixieBrix runtime.
 * @see runStage
 * @see reducePipeline
 */
export type BrickConfig = {
  /**
   * The registry id of the configured brick
   */
  id: RegistryId;

  /**
   * The configuration of the brick
   */
  // XXX: in the future, should use JsonObject here, but will require extra type assertions in the codebase.
  config: UnknownObject;

  /**
   * (Optional) human-readable label for the step.
   *
   * Shown in the page editor, logs, and the progress indicator
   * @see notifyProgress
   */
  label?: string;

  /**
   * (Optional) `true` to indicate on the host page that the step is being run
   */
  notifyProgress?: boolean;

  /**
   * (Optional) If the brick is being run in the context of a deployment, if an error occurs in this step an
   * alert email is sent to the admins for the deployment with the brick input.
   *
   * Used as a stopgap/recovery mechanism operations that aren't safe to retry. (For example, sending data to a
   * UiPath queue for execution)
   */
  onError?: {
    alert?: boolean;
  };

  /**
   * Where to execute the brick (default=`self`)
   * @see BrickWindow
   */
  window?: BrickWindow;

  /**
   * The output key (without the preceding "@") to assign the brick output to
   *
   * As of brick API `v2`, the outputKey is required except:
   * - Effect bricks (because they don't produce an output)
   * - The last brick in pipeline
   *
   * @see ApiVersionOptions.explicitDataFlow
   */
  outputKey?: OutputKey;

  /**
   * (Optional) condition expression written in templateEngine for deciding if the step should be run. If not
   * provided, the step is run unconditionally.
   * @see BlockIf
   */
  if?: BlockIf;

  /**
   * (Optional) whether the brick should inherit the current root element, or if it should use the document
   * (default=`inherit`)
   *
   * `element` rootMode added in 1.7.16
   *
   * @see root
   * @since 1.4.0
   */
  rootMode?: "inherit" | "document" | "element";

  /**
   * (Optional) root jQuery/CSS selector or element reference.
   *
   * For rootMode of "inherit" and "document", interpreted as a jQuery/CSS selector. The selector is relative to the
   * `root` that is passed to the pipeline/stage. An error is thrown at runtime if the selector doesn't match exactly
   * one element
   *
   * @see rootMode
   */
  root?: string | ElementReference | Expression;

  /**
   * (Optional) template language to use for rendering the `if` and `config` properties (default=`mustache`)
   * @see config
   * @see if
   */
  templateEngine?: TemplateEngine;

  /**
   * A unique id for the configured brick, used to correlate traces across runs when using the Page Editor.
   *
   * DO NOT SET: generated automatically by the Page Editor when configuring a dynamic element.
   */
  instanceId?: UUID;
};

/**
 * A pipeline of bricks to execute sequentially
 */
export type BrickPipeline = BrickConfig[];

/**
 * A control flow branch, for tracing. The array of branches is used to correlate runs of the same brick in a pipeline
 * @see TraceMetadata.blockInstanceId
 * @see TraceMetadata.branches
 */
export type Branch = {
  key: string;
  counter: number;
};

/**
 * Defines the position of the brick in the extension
 * ex. "extension.brickPipeline.0.config.body.__value__.0",
 * "extension.brickPipeline.0.config.body.0.children.0.config.onClick.__value__.0"
 */
export type BrickPosition = {
  /**
   * The path to the brick relative to the root pipeline
   */
  path: string;
};
