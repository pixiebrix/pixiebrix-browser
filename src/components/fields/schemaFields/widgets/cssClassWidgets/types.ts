/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { type Expression } from "@/types/runtimeTypes";
import type React from "react";

export type Spacing = {
  side: string | null;
  size: number;
};

export type Value = string | Expression;

/**
 * An independent class name
 */
export type ClassFlag = {
  /**
   * The Bootstrap 4 class name
   */
  className: string;

  /**
   * Title node to render for the element (in a button/dropdown)
   */
  title: React.ReactNode;

  /**
   * True if the flag is exclusive for it's group (default=true)
   */
  exclusive?: boolean;

  /**
   * Other flags in the same group that the flag implies
   */
  implies?: string[];
};
