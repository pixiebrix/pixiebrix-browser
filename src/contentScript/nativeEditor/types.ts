/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { type Framework } from "@/messaging/constants";
import {
  type ExtensionPointConfig,
  type ExtensionPointDefinition,
  type ExtensionPointType,
} from "@/extensionPoints/types";
import { type EmptyConfig, type IExtension, type UUID } from "@/core";
import { type Except } from "type-fest";
import {
  type PanelConfig,
  type PanelDefinition,
} from "@/extensionPoints/panelExtension";
import {
  type MenuDefinition,
  type MenuItemExtensionConfig,
} from "@/extensionPoints/menuItemExtension";

export interface ElementInfo {
  selectors: string[];
  framework: Framework;
  tagName: string;
  hasData: boolean;
  parent?: ElementInfo;
  isMulti?: boolean;
}

export interface DynamicDefinition<
  TExtensionPoint extends ExtensionPointDefinition = ExtensionPointDefinition,
  TExtension extends EmptyConfig = EmptyConfig
> {
  type: ExtensionPointType;
  extensionPoint: ExtensionPointConfig<TExtensionPoint>;
  extension: IExtension<TExtension>;
}

export type SelectMode = "element" | "container";
export type PanelSelectionResult = {
  uuid: UUID;
  foundation: Except<
    PanelDefinition,
    "defaultOptions" | "isAvailable" | "reader"
  >;
  panel: Except<PanelConfig, "body">;
  containerInfo: ElementInfo;
};
export type ButtonDefinition = DynamicDefinition<
  MenuDefinition,
  MenuItemExtensionConfig
>;
export type ButtonSelectionResult = {
  uuid: UUID;
  menu: Except<MenuDefinition, "defaultOptions" | "isAvailable" | "reader">;
  item: Pick<MenuItemExtensionConfig, "caption">;
  containerInfo: ElementInfo;
};
