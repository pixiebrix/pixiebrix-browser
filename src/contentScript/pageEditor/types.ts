/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import {
  type StarterBrickConfig,
  type StarterBrickDefinition,
  type StarterBrickType,
} from "@/starterBricks/types";
import { type Except } from "type-fest";
import {
  type PanelConfig,
  type PanelDefinition,
} from "@/starterBricks/panelExtension";
import {
  type MenuDefinition,
  type MenuItemStarterBrickConfig,
} from "@/starterBricks/menuItemExtension";
import { type ElementInfo } from "@/pageScript/frameworks";
import { type ModComponentBase } from "@/types/modComponentTypes";
import { type UnknownObject } from "@/types/objectTypes";
import { type UUID } from "@/types/stringTypes";

export interface DynamicDefinition<
  TExtensionPoint extends StarterBrickDefinition = StarterBrickDefinition,
  TExtension extends UnknownObject = UnknownObject
> {
  type: StarterBrickType;
  extensionPointConfig: StarterBrickConfig<TExtensionPoint>;
  extension: ModComponentBase<TExtension>;
}

export type SelectMode = "element" | "container";
export type PanelSelectionResult = {
  uuid: UUID;
  foundation: Except<
    PanelDefinition,
    "defaultOptions" | "isAvailable" | "reader"
  >;
  panel: Except<PanelConfig, "body">;
  containerInfo: ElementInfo;
};
export type ButtonDefinition = DynamicDefinition<
  MenuDefinition,
  MenuItemStarterBrickConfig
>;
export type ButtonSelectionResult = {
  uuid: UUID;
  menu: Except<MenuDefinition, "defaultOptions" | "isAvailable" | "reader">;
  item: Pick<MenuItemStarterBrickConfig, "caption">;
  containerInfo: ElementInfo;
};

export type AttributeExample = {
  name: string;
  value: string;
};
