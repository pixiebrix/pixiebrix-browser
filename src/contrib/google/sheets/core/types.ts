/* eslint-disable @typescript-eslint/no-redundant-type-constituents -- TODO: The rule is right, but is this used for documentation? */
/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

export interface SheetMeta {
  id: string;
  name: string;
}

export interface Doc extends SheetMeta {
  serviceId: "spread" | string;
  mimeType: "application/vnd.google-apps.spreadsheet" | string;
}

export interface Data {
  action: string;
  docs: Doc[];
}

// GSheets API Proxy types
export type ValueRange = gapi.client.sheets.ValueRange;
export type AppendValuesResponse = gapi.client.sheets.AppendValuesResponse;
export type BatchUpdateSpreadsheetRequest =
  gapi.client.sheets.BatchUpdateSpreadsheetRequest;
export type BatchUpdateSpreadsheetResponse =
  gapi.client.sheets.BatchUpdateSpreadsheetResponse;
export type Spreadsheet = gapi.client.sheets.Spreadsheet;
export type SpreadsheetProperties = gapi.client.sheets.SpreadsheetProperties;

export type File = {
  kind: "drive#file";
  mimeType: string;
  id: string;
  name: string;
};
export type FileList = {
  kind: "drive#fileList";
  incompleteSearch: boolean;
  files: File[];
  nextPageToken?: string;
};
