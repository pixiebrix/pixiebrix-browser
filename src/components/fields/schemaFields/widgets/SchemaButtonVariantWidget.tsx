/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useMemo } from "react";
import { type SchemaFieldProps } from "@/components/fields/schemaFields/propTypes";
import { useField } from "formik";
import { type Option } from "@/components/form/widgets/SelectWidget";
import Select, {
  type OptionProps,
  type ValueContainerProps,
} from "react-select";
import { Button } from "react-bootstrap";
import styles from "./SchemaButtonVariantWidget.module.scss";
import cx from "classnames";
import { type SingleValueProps } from "react-select/dist/declarations/src/components/SingleValue";
import { type StringOption } from "@/components/fields/schemaFields/widgets/SchemaSelectWidget";
import { mapSchemaToOptions } from "@/components/fields/schemaFields/selectFieldUtils";

const OptionComponent = ({
  innerProps,
  innerRef,
  data,
  isSelected,
}: OptionProps<Option>) => (
  <div
    ref={innerRef}
    {...innerProps}
    className={cx(styles.optionContainer, { [styles.active]: isSelected })}
  >
    <Button
      data-testid="variant-option"
      type={null}
      variant={data.value}
      size="sm"
      className={styles.exampleButton}
    >
      {data.label}
    </Button>
  </div>
);

const ValueComponent = ({ data }: SingleValueProps<Option>) => (
  <Button
    data-testid="selected-variant"
    type={null}
    variant={data.value}
    size="sm"
    className={styles.exampleButton}
  >
    {data.label}
  </Button>
);

const ContainerComponent = ({
  innerProps,
  children,
}: ValueContainerProps<Option>) => (
  <div {...innerProps} className={styles.selectContainer}>
    {children}
  </div>
);

const SchemaButtonVariantWidget: React.FunctionComponent<SchemaFieldProps> = ({
  name,
  schema,
}) => {
  const [{ value }, , { setValue }] = useField(name);
  const { options } = useMemo(
    () => mapSchemaToOptions({ schema, value }),
    [schema, value]
  );

  const selectedValue = options.find((x) => x.value === value) ?? {
    value: null,
  };

  return (
    <div className="mt-2" data-testid="select-container">
      <Select
        name={name}
        options={options}
        value={selectedValue}
        isSearchable={false}
        // Set clearable to false because the input looks off next to the button preview
        isClearable={false}
        onChange={(event: StringOption) => {
          setValue(event.value);
        }}
        components={{
          Option: OptionComponent,
          SingleValue: ValueComponent,
          ValueContainer: ContainerComponent,
        }}
      />
    </div>
  );
};

export default SchemaButtonVariantWidget;
