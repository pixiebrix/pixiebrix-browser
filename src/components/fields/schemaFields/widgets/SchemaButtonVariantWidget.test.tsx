/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import SchemaButtonVariantWidget from "@/components/fields/schemaFields/widgets/SchemaButtonVariantWidget";
import { render } from "@/pageEditor/testHelpers";
import { type Schema } from "@/types/schemaTypes";
// eslint-disable-next-line no-restricted-imports
import { Formik } from "formik";
import selectEvent from "react-select-event";
import userEvent from "@testing-library/user-event";

const fieldName = "testField";
const fieldDescription = "this is a test field description";

const schema: Schema = {
  type: "string",
  oneOf: [
    { const: "primary", title: "Primary" },
    { const: "outline-primary", title: "Primary" },
    { const: "secondary", title: "Secondary" },
    { const: "outline-secondary", title: "Secondary" },
    { const: "success", title: "Success" },
    { const: "outline-success", title: "Success" },
    { const: "warning", title: "Warning" },
    { const: "outline-warning", title: "Warning" },
    { const: "danger", title: "Danger" },
    { const: "outline-danger", title: "Danger" },
  ],
};

const renderSelect = (value: string, onSubmit?: (formVals: any) => void) =>
  render(
    <Formik
      initialValues={{ testField: value }}
      onSubmit={onSubmit}
      validate={onSubmit}
      validateOnChange
    >
      <SchemaButtonVariantWidget
        name={fieldName}
        schema={schema}
        description={fieldDescription}
      />
    </Formik>
  );

describe("SchemaButtonVariantWidget", () => {
  test("renders button variant select widget", () => {
    const defaultSelect = renderSelect("outline-primary");

    // General snapshot test
    expect(defaultSelect).toMatchSnapshot();

    // Selected variant matches preview
    const { getByTestId } = defaultSelect;
    expect(getByTestId("selected-variant")).toHaveClass("btn-outline-primary");
  });

  test("selecting variants updates form", async () => {
    const onSubmit = jest.fn();

    const defaultSelect = renderSelect("outline-primary", (values) =>
      onSubmit(values)
    );
    const { getByTestId, queryAllByTestId } = defaultSelect;

    // All variants present
    const selectContainerElement =
      getByTestId("select-container").querySelector("div");

    selectEvent.openMenu(selectContainerElement);
    const options = queryAllByTestId("variant-option");
    expect(options).toHaveLength(10);

    await userEvent.click(options[4]);

    expect(getByTestId("selected-variant")).toHaveClass("btn-success");
    expect(onSubmit).toHaveBeenCalledWith({ testField: "success" });
  });
});
