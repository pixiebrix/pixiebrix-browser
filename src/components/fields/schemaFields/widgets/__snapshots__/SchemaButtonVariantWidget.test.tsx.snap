// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaButtonVariantWidget renders button variant select widget 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mt-2"
        data-testid="select-container"
      >
        <div
          class=" css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-2-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
          />
          <div
            class=" css-13cymwt-control"
          >
            <div
              class="selectContainer"
            >
              <button
                class="exampleButton btn btn-outline-primary btn-sm"
                data-testid="selected-variant"
                type="button"
              >
                Primary
              </button>
              <input
                aria-autocomplete="list"
                aria-expanded="false"
                aria-haspopup="true"
                aria-readonly="true"
                class="css-mohuvp-dummyInput-DummyInput"
                id="react-select-2-input"
                inputmode="none"
                role="combobox"
                tabindex="0"
                value=""
              />
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class=" css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <input
            name="testField"
            type="hidden"
            value="outline-primary"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mt-2"
      data-testid="select-container"
    >
      <div
        class=" css-b62m3t-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-13cymwt-control"
        >
          <div
            class="selectContainer"
          >
            <button
              class="exampleButton btn btn-outline-primary btn-sm"
              data-testid="selected-variant"
              type="button"
            >
              Primary
            </button>
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-haspopup="true"
              aria-readonly="true"
              class="css-mohuvp-dummyInput-DummyInput"
              id="react-select-2-input"
              inputmode="none"
              role="combobox"
              tabindex="0"
              value=""
            />
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class=" css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="testField"
          type="hidden"
          value="outline-primary"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getFormState": [Function],
  "getReduxStore": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
  "updateFormState": [Function],
}
`;
